pyro/params/param_store.py

```diff
@@ -295,6 +299,44 @@ def load(self, filename, map_location=None):
             state = torch.load(input_file, map_location)
         self.set_state(state)
 
+    @contextmanager
+    def scope(self, state=None) -> dict:
+        """
+        Context manager for using multiple parameter stores within the same process.
+
+        This is a thin wrapper around :meth:`get_state`, :meth:`clear`, and
+        :meth:`set_state`. For large models where memory space is limiting, you
+        may want to instead manually use :meth:`save`, :meth:`clear`, and
+        :meth:`load`.
+
+        Example usage::
+
+            param_store = pyro.get_param_store()
+
+            # Train multiple models, while avoiding param name conflicts.
+            with param_store.scope() as scope1:
+                # ...Train one model,guide pair...
+            with param_store.scope() as scope2:
+                # ...Train another model,guide pair...
+
+            # Now evaluate each, still avoiding name conflicts.
+            with param_store.scope(scope1):  # loads the first model's scope
+               # ...evaluate the first model...
+            with param_store.scope(scope2):  # loads the second model's scope
+               # ...evaluate the second model...
+        """
+        if state is None:
+            state = {"params": {}, "constraints": {}}
+        old_state = self.get_state()
```

Comment:
Is there merit in raising an exception when scopes are nested? Or are nested scopes a valid use case?

Reply:
I think nested scopes are valid use cases